llama_index/node_parser/file/markdown.py

```diff
@@ -0,0 +1,146 @@
+        include_prev_next_rel (bool): whether to include prev/next relationships
+
+    """
+
+    include_metadata: bool = Field(
+        default=True, description="Whether or not to consider metadata when splitting."
+    )
+    include_prev_next_rel: bool = Field(
+        default=True, description="Include prev/next node relationships."
+    )
+    callback_manager: CallbackManager = Field(
+        default_factory=CallbackManager, exclude=True
+    )
+
+    @classmethod
+    def from_defaults(
+        cls,
+        include_metadata: bool = True,
+        include_prev_next_rel: bool = True,
+        callback_manager: Optional[CallbackManager] = None,
+    ) -> "MarkdownNodeParser":
+        callback_manager = callback_manager or CallbackManager([])
+
+        return cls(
+            include_metadata=include_metadata,
+            include_prev_next_rel=include_prev_next_rel,
+            callback_manager=callback_manager,
+        )
+
+    @classmethod
+    def class_name(cls) -> str:
+        """Get class name."""
+        return "MarkdownNodeParser"
+
+    def get_nodes_from_documents(
+        self,
+        documents: Sequence[TextNode],
+        show_progress: bool = False,
+    ) -> List[BaseNode]:
+        """Parse document into nodes.
+
+        Args:
+            documents (Sequence[TextNode]): TextNodes or Documents to parse
+
+        """
+        with self.callback_manager.event(
+            CBEventType.NODE_PARSING, payload={EventPayload.DOCUMENTS: documents}
+        ) as event:
+            all_nodes: List[BaseNode] = []
+            documents_with_progress = get_tqdm_iterable(
+                documents, show_progress, "Parsing documents into nodes"
+            )
+
+            for document in documents_with_progress:
+                nodes = self.get_nodes_from_node(document)
+                all_nodes.extend(nodes)
+
+            event.on_end(payload={EventPayload.NODES: all_nodes})
+
+        return all_nodes
+
+    def get_nodes_from_node(self, node: BaseNode) -> List[TextNode]:
+        """Get nodes from document"""
+
+        text = node.get_content(metadata_mode=MetadataMode.NONE)
+        markdown_nodes = []
+        lines = text.split("\n")
+        metadata: dict[str, str] = {}
+
+        current_section = ""
+
+        for line in lines:
+            header_match = re.match(r"^(#+)\s(.*)", line)
+            if header_match:
+                if current_section != "":
+                    markdown_nodes.append(
+                        self._build_node_from_split(
+                            current_section.strip(), node, metadata
+                        )
+                    )
+                metadata = self._update_metadata(
+                    metadata, header_match.group(2), len(header_match.group(1).strip())
+                )
+                current_section = f"{header_match.group(2)}\n"
+            else:
+                current_section += line + "\n"
```

Comment:
is the `\n` needed? Does the line already have a `\n` at the end?

Reply:
Since we split on `\n` we add it back here